### PR TITLE
Add forwards compat tests

### PR DESCRIPTION
**Pull request by jkotalik**
@pakrym I think I may add a few more properties to skip different tests (lot of duplication right now). Besides that, I think this is ready for review.

### CODE REVIEW

File: .vsts-pipelines/templates/build-steps.yml [hunk 1]
```diff
@@ -43,7 +43,7 @@ phases:
   parameters:
     agentOs: Windows
     phaseName: IIS_BackCompat_Tests
-    buildArgs: /p:SkipIISTests=true /p:SkipIISExpressTests=true
+    buildArgs: /p:SkipIISTests=true /p:SkipIISExpressTests=true /p:SkipForwardsCompatTests
```
Comment (pakrym) [inline .vsts-pipelines/templates/build-steps.yml]: you need true here.

File: .vsts-pipelines/templates/build-steps.yml [hunk 2]
```diff
@@ -59,6 +59,26 @@ phases:
           artifactType: Container
           pathtoPublish: artifacts/logs
 
+          - template: .vsts-pipelines/templates/phases/default-build.yml@buildtools
```
Comment (pakrym) [inline .vsts-pipelines/templates/build-steps.yml]: Formatting here is off

File: test/IIS.ForwardsCompatibility.FunctionalTests/RequiresNewComponent.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using System;
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Testing.xunit;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method)]
+    public sealed class RequiresNewShimAttribute : Attribute, ITestCondition
+    {
+        public bool IsMet => true;
```
Comment (pakrym) [inline test/IIS.ForwardsCompatibility.FunctionalTests/RequiresNewComponent.cs]: Now I'm thinking about keeping one copy of every attribute and grabbing info from DeployerSelector in it.

File: test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (pakrym) [inline test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj]: Rename this project file.

File: test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    <DisableFastUpToDateCheck>True</DisableFastUpToDateCheck>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Content Include="..\Common.FunctionalTests\AppHostConfig\*.config" CopyToOutputDirectory="PreserveNewest" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Common.Tests\Common.Tests.csproj" />
+    <ProjectReference Include="..\WebSites\InProcessWebSite\InProcessWebSite.csproj">
```
Comment (pakrym) [inline test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: You are referencing the wrong site. Also, do you need all project references here?

File: test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+  <Import Project="..\..\..\build\testsite.props" />
+
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp2.2</TargetFrameworks>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="..\shared\**\*.cs" />
+    <Compile Include="..\InProcessWebSite\*.cs" />
```
Comment (pakrym) [inline test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj]: Use a recursive wildcard (`**\*.cs`)

File: test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj (hunk 4 above)
Comment (jkotalik) [reply test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj]: I did this on purpose. We have ~3 or 4 scenarios that depend on the InProcessWebSite name for testing (runtimeconfig.json, exe checks, etc)

File: test/Common.FunctionalTests/Utilities/Helpers.cs [hunk 7]
```diff
@@ -27,7 +27,10 @@ public static string GetTestWebSitePath(string name)
             return Path.Combine(TestPathUtilities.GetSolutionRootDirectory("IISIntegration"),"test", "WebSites", name);
         }
 
-        public static string GetInProcessTestSitesPath() => GetTestWebSitePath("InProcessWebSite");
+        public static string GetInProcessTestSitesPath()
+        {
+            return DeployerSelector.IsForwardsCompatibilityTest ? GetTestWebSitePath("InProcessForwardsCompatWebSite") : GetTestWebSitePath("InProcessWebSite");
```
Comment (pakrym) [inline test/Common.FunctionalTests/Utilities/Helpers.cs]: Expose InProcessSiteNam in DeployerSelector.

File: test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="$(MicrosoftExtensionsLoggingPackageVersion)" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Testing" Version="$(MicrosoftExtensionsLoggingTestingPackageVersion)" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkPackageVersion)" />
+    <PackageReference Include="System.Diagnostics.EventLog" Version="$(SystemDiagnosticsEventLogPackageVersion)" />
+    <PackageReference Include="System.Net.WebSockets.WebSocketProtocol" Version="$(SystemNetWebSocketsWebSocketProtocolPackageVersion)" />
+    <PackageReference Include="xunit" Version="$(XunitPackageVersion)" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="$(XunitRunnerVisualStudioPackageVersion)" />
+  </ItemGroup>
+
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: No, we can drop all of them.

File: test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj (hunk 4 above)
Comment (pakrym) [reply test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj]: Hmmmmm, fine.

File: test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj (hunk 5 above)
Comment (pakrym) [reply test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: Incremental compilation

**Comment by pakrym** [comment]
Add a test that verifies that we use old handler

File: test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj (hunk 6 above)
Comment (jkotalik) [reply test/WebSites/InProcessForwardsCompatWebSite/InProcessWebSite.csproj]: It will start pulling in obj output from the InProcessWebSite.

**Comment by pakrym** [comment]
Use IISTestSiteFixture instead of deploying in test

**Review by pakrym — APPROVED** [review]

**Comment by jkotalik** [comment]
Test failure https://github.com/aspnet/IISIntegration/issues/1447

**Pull request merged**